prune.go

```diff
@@ -129,122 +135,194 @@ func txnsPruneC(txnsName string) string {
-	logger.Debugf("%d transactions to remove", len(txnIds))
-	if err := bulkRemoveTxns(txns, txnIds); err != nil {
-		return fmt.Errorf("txn removal failed: %v", err)
+	// Remove the no-longer-referenced transactions from the txns collection.
+	t = newSimpleTimer(logInterval)
+	remover = newBulkRemover(txns)
+	defer remover.close()
+	query := workingSet.Find(nil).Batch(batchBytes)
+	iter := query.Iter()
+	var doc bson.D
+	for iter.Next(&doc) {
+		if err := remover.remove(doc[0].Value); err != nil {
+			return fmt.Errorf("removing txns: %v", err)
+		}
+		if t.isAfter() {
+			logger.Debugf("%d completed txns pruned so far", remover.removed)
+		}
+	}
+	if err := remover.flush(); err != nil {
+		return fmt.Errorf("removing txns: %v", err)
 	}
-	logger.Debugf("completed transactions pruned")
+
+	logger.Debugf("pruning completed: removed %d txns", remover.removed)
 	return nil
 }
 
+func runPipe(pipe *mgo.Pipe, result interface{}) error {
+	pipe.Batch(batchBytes)
+	pipe.AllowDiskUse()
+	return pipe.All(result)
+}
+
 func txnTokenToId(token string) bson.ObjectId {
 	// mgo/txn transaction tokens are the 24 character txn id
 	// followed by "_<nonce>"
 	return bson.ObjectIdHex(token[:24])
 }
 
-// bulkRemoveTxns removes transaction documents in chunks. It should
-// be significantly more efficient than removing one document per
-// remove query while also not trigger query document size limits.
-func bulkRemoveTxns(txns *mgo.Collection, txnIds map[bson.ObjectId]struct{}) error {
-	removeCount := 0
-	removeTxns := func(ids []bson.ObjectId) error {
-		bulk := txns.Bulk()
-		bulk.Unordered()
-		for _, id := range ids {
-			bulk.Remove(bson.D{{"_id", id}})
-		}
-		switch _, err := bulk.Run(); err {
-		case nil, mgo.ErrNotFound:
-			// It's OK for txns to no longer exist. Another process
-			// may have concurrently pruned them.
-			removeCount += len(ids)
-			logger.Tracef("%d completed transactions removed", removeCount)
-			return nil
-		default:
-			return err
-		}
+func newBulkRemover(coll *mgo.Collection) *bulkRemover {
+	session := coll.Database.Session.Copy()
+	r := &bulkRemover{
+		coll: coll.With(session),
 	}
+	r.newChunk()
+	return r
+}
 
-	// There must not be more than 1000 operations in a bulk operation.
-	const chunkSize = 1000
-	chunk := make([]bson.ObjectId, 0, chunkSize)
-	for txnId := range txnIds {
-		chunk = append(chunk, txnId)
-		if len(chunk) == chunkSize {
-			if err := removeTxns(chunk); err != nil {
-				return err
-			}
-			chunk = chunk[:0] // Avoid reallocation.
-		}
+type bulkRemover struct {
+	coll      *mgo.Collection
+	chunk     *mgo.Bulk
+	chunkSize int
+	removed   int
+}
+
+// There must not be more than 1000 operations in a bulk operation.
+const maxChunkSize = 1000
+
+func (r *bulkRemover) newChunk() {
+	r.chunk = r.coll.Bulk()
+	r.chunk.Unordered()
+	r.chunkSize = 0
+}
+
+func (r *bulkRemover) remove(id interface{}) error {
+	r.chunk.Remove(bson.D{{"_id", id}})
+	r.chunkSize++
+	if r.chunkSize == maxChunkSize {
```

Comment:
As a general peeve of mine, I *really* prefer these checks to be:
```
if r.chunkSize >= maxChunkSize {
}
```

its a bit of belt-and-suspenders checking, but "==" always has the catastrophic failure when you accidentally incremented twice and suddenly the trap never triggers because you walked right past it.